### PR TITLE
SQL: Use exact attributes for script templates from scalar functions

### LINKED ISSUE #80551: SQL: Group by expression with text field does not always use `.keyword` subfield
Given the mapping

```
       column       |     type      |    mapping    
--------------------+---------------+---------------
...
last_name           |VARCHAR        |text           
last_name.keyword   |VARCHAR        |keyword        
...
```

the query `SELECT last_name IS NULL FROM test_emp GROUP BY 1` fails with 

```
{
  "error": {
    "root_cause": [
      {
        "type": "script_exception",
        "reason": "runtime error",
        "script_stack": [
          "org.elasticsearch.index.mapper.TextFieldMapper$TextFieldType.fielddataBuilder(TextFieldMapper.java:814)",
          "org.elasticsearch.index.fielddata.IndexFieldDataService.getForField(IndexFieldDataService.java:109)",
          "org.elasticsearch.index.query.SearchExecutionContext.lambda$lookup$2(SearchExecutionContext.java:503)",
          "org.elasticsearch.search.lookup.SearchLookup.getForField(SearchLookup.java:105)",
          "org.elasticsearch.search.lookup.LeafDocLookup$1.run(LeafDocLookup.java:72)",
          "org.elasticsearch.search.lookup.LeafDocLookup$1.run(LeafDocLookup.java:69)",
          "java.base/java.security.AccessController.doPrivileged(AccessController.java:318)",
          "org.elasticsearch.search.lookup.LeafDocLookup.get(LeafDocLookup.java:69)",
          "org.elasticsearch.search.lookup.LeafDocLookup.get(LeafDocLookup.java:27)",
          "org.elasticsearch.xpack.ql.expression.function.scalar.whitelist.InternalQlScriptUtils.docValue(InternalQlScriptUtils.java:37)",
          "InternalQlScriptUtils.isNull(InternalQlScriptUtils.docValue(doc,params.v0))",
          "                                                                      ^---- HERE"
        ],
        "script": "InternalQlScriptUtils.isNull(InternalQlScriptUtils.docValue(doc,params.v0))",
        "lang": "painless",
        "position": {
          "offset": 70,
          "start": 0,
          "end": 75
        }
      }
    ],
    "type": "search_phase_execution_exception",
    "reason": "all shards failed",
    "phase": "query",
    "grouped": true,
    "failed_shards": [
      {
        "shard": 0,
        "index": "test_emp",
        "node": "6a6CBLYYQFa6gqsxbMkn5Q",
        "reason": {
          "type": "script_exception",
          "reason": "runtime error",
          "script_stack": [
            "org.elasticsearch.index.mapper.TextFieldMapper$TextFieldType.fielddataBuilder(TextFieldMapper.java:814)",
            "org.elasticsearch.index.fielddata.IndexFieldDataService.getForField(IndexFieldDataService.java:109)",
            "org.elasticsearch.index.query.SearchExecutionContext.lambda$lookup$2(SearchExecutionContext.java:503)",
            "org.elasticsearch.search.lookup.SearchLookup.getForField(SearchLookup.java:105)",
            "org.elasticsearch.search.lookup.LeafDocLookup$1.run(LeafDocLookup.java:72)",
            "org.elasticsearch.search.lookup.LeafDocLookup$1.run(LeafDocLookup.java:69)",
            "java.base/java.security.AccessController.doPrivileged(AccessController.java:318)",
            "org.elasticsearch.search.lookup.LeafDocLookup.get(LeafDocLookup.java:69)",
            "org.elasticsearch.search.lookup.LeafDocLookup.get(LeafDocLookup.java:27)",
            "org.elasticsearch.xpack.ql.expression.function.scalar.whitelist.InternalQlScriptUtils.docValue(InternalQlScriptUtils.java:37)",
            "InternalQlScriptUtils.isNull(InternalQlScriptUtils.docValue(doc,params.v0))",
            "                                                                      ^---- HERE"
          ],
          "script": "InternalQlScriptUtils.isNull(InternalQlScriptUtils.docValue(doc,params.v0))",
          "lang": "painless",
          "position": {
            "offset": 70,
            "start": 0,
            "end": 75
          },
          "caused_by": {
            "type": "illegal_argument_exception",
            "reason": "Text fields are not optimised for operations that require per-document field data like aggregations and sorting, so these operations are disabled by default. Please use a keyword field instead. Alternatively, set fielddata=true on [last_name] in order to load field data by uninverting the inverted index. Note that this can use significant memory."
          }
        }
      }
    ]
  },
  "status": 400
}
```

`/translate` reveals that SQL does not use the `last_name.keyword` subfield in this case:

```
{
  "size": 0,
  "_source": false,
  "aggregations": {
    "groupby": {
      "composite": {
        "size": 1000,
        "sources": [
          {
            "4642bd56": {
              "terms": {
                "script": {
                  "source": "InternalQlScriptUtils.isNull(InternalQlScriptUtils.docValue(doc,params.v0))",
                  "lang": "painless",
                  "params": {
                    "v0": "last_name"
                  }
                },
                "missing_bucket": true,
                "value_type": "boolean",
                "order": "asc"
              }
            }
          }
        ]
      }
    }
  }
}
```

Explicitly using the `.keyword` subfield works (`SELECT last_name.keyword IS NULL FROM test_emp GROUP BY 1`).

Also some other queries like `SELECT CONCAT('Mr. ', last_name) FROM test_emp GROUP BY 1` work as expected.

It looks like only queries grouping on conditions that can be expressed as ES filters are affected by this issue.

### PR DESCRIPTION
**Pull request by luigidellaquila**
Fixes #80551

This fix affects both SQL and EQL

### CODE REVIEW

**Comment by luigidellaquila** [comment]
@elasticmachine update branch

**Review by costin — APPROVED** [review]
LGTM

**Review by bpintea — APPROVED** [review]
LGTM

File: x-pack/plugin/eql/src/test/resources/querytranslator_tests.txt [hunk 1]
```diff
@@ -841,7 +841,7 @@ disjunctionOfMultiFieldsFunctions
 process where concat(file_name, ".", process_name) == "foo" or add(pid, ppid) > 100
 ;
 {"bool":{"should":[{"script":{"script":{"source":"InternalEqlScriptUtils.multiValueDocValues(doc,params.v0,X0->InternalEqlScriptUtils.multiValueDocValues(doc,params.v1,X1->InternalQlScriptUtils.nullSafeFilter(InternalQlScriptUtils.eq(InternalEqlScriptUtils.concat([X0,params.v2,X1]),params.v3))))"
-"params":{"v0":"file_name","v1":"process_name","v2":".","v3":"foo"}}
+"params":{"v0":"file_name.keyword","v1":"process_name","v2":".","v3":"foo"}}
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/resources/querytranslator_tests.txt:844]: This change raises some questions:
- why `v0` has this change while `v1` doesn't? Both are parameters of `InternalEqlScriptUtils.multiValueDocValues` script method
- what field types are `file_name` and `process_name`? Is one `text` (the `file_name`) and the other `keyword` (the `process_name`) and this explains why the change applies to only one of them?
- how come no one complained about this behavior so far? The multi-value behavior is the default for few versions now, and I was expecting this to hit at least one user. Is everyone using just `keyword` fields or is there some other explanation to why this worked fine so far?

**Review by astefan — COMMENTED** [review]
The PR LGTM but I have one concern/question that needs an explanation, imo, regarding a specific change in EQL.

File: x-pack/plugin/eql/src/test/resources/querytranslator_tests.txt (hunk 1 above)
Comment (luigidellaquila) [reply x-pack/plugin/eql/src/test/resources/querytranslator_tests.txt:844]: Your guess is correct, `file_name` is `text`, while `process_name` is `keyword` (see https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/eql/src/test/resources/mapping-default.json)

About last question, unfortunately I don't have an answer. My guess is that, even though many operators are affected, the actual use cases are quite uncommon (eg. equality in the projections or usage of scalar functions in GROUP BY). Another explanation could be that the work-around is easy to guess, eg. a query like 

```
SELECT LEAST(a, b) x FROM ...
```

can easily be rewritten as 

```
SELECT LEAST(a.keyword, b.keyword) x FROM ...
```

so probably it's easier for the end user to adapt to the limitation than to file a new issue

**Review by Luegg — APPROVED** [review]
Lgtm. I would also mention in the PR description that EQL is affected as well. Can we have more than one `area` label in the changelog?

**Comment by luigidellaquila** [comment]
Thanks @Luegg , good point.
About the labels, let me just try and see if the bots generates the yaml accordingly

**Comment by luigidellaquila** [comment]
No, it seems the changelog was not updated. Checking the old changelog entries, I didn't find any with two areas, so I guess it's not possible.

**Review by astefan — APPROVED** [review]
LGTM 👍

**Pull request merged**